ros/src/computing/planning/motion/packages/op_local_planner/nodes/op_behavior_selector/op_behavior_selector_core.cpp

```diff
@@ -0,0 +1,495 @@
+	bRollOuts = false;
+
+
+	ros::NodeHandle _nh;
+	UpdatePlanningParams(_nh);
+
+	tf::StampedTransform transform;
+	PlannerHNS::RosHelpers::GetTransformFromTF("map", "world", transform);
+	m_OriginPos.position.x  = transform.getOrigin().x();
+	m_OriginPos.position.y  = transform.getOrigin().y();
+	m_OriginPos.position.z  = transform.getOrigin().z();
+
+	pub_LocalPath = nh.advertise<autoware_msgs::lane>("final_waypoints", 1,true);
+	pub_LocalBasePath = nh.advertise<autoware_msgs::lane>("base_waypoints", 1,true);
+	pub_ClosestIndex = nh.advertise<std_msgs::Int32>("closest_waypoint", 1,true);
+	pub_BehaviorState = nh.advertise<geometry_msgs::TwistStamped>("current_behavior", 1);
+	pub_SimuBoxPose	  = nh.advertise<geometry_msgs::PoseArray>("sim_box_pose_ego", 1);
+	pub_BehaviorStateRviz = nh.advertise<visualization_msgs::Marker>("behavior_state", 1);
+
+	sub_current_pose 	= nh.subscribe("/current_pose", 			1,		&BehaviorGen::callbackGetCurrentPose, 		this);
+
+	int bVelSource = 1;
+	_nh.getParam("/op_trajectory_evaluator/velocitySource", bVelSource);
+	if(bVelSource == 0)
+		sub_robot_odom 			= nh.subscribe("/odom", 					100,	&BehaviorGen::callbackGetRobotOdom, 	this);
+	else if(bVelSource == 1)
+		sub_current_velocity 	= nh.subscribe("/current_velocity",		100,	&BehaviorGen::callbackGetVehicleStatus, 	this);
+	else if(bVelSource == 2)
+		sub_can_info 			= nh.subscribe("/can_info",		100,	&BehaviorGen::callbackGetCanInfo, 	this);
+
+	sub_GlobalPlannerPaths 	= nh.subscribe("/lane_waypoints_array", 	1,		&BehaviorGen::callbackGetGlobalPlannerPath, 	this);
+	sub_LocalPlannerPaths 	= nh.subscribe("/local_weighted_trajectories", 	1,		&BehaviorGen::callbackGetLocalPlannerPath, 	this);
+	sub_TrafficLightStatus 		= nh.subscribe("/light_color", 		1,		&BehaviorGen::callbackGetTrafficLightStatus, 	this);
+	sub_TrafficLightSignals		= nh.subscribe("/roi_signal", 		1,		&BehaviorGen::callbackGetTrafficLightSignals, 	this);
+	sub_Trajectory_Cost			=  nh.subscribe("/local_trajectory_cost", 		1,		&BehaviorGen::callbackGetLocalTrajectoryCost, 	this);
+
+	sub_twist_raw = nh.subscribe("/twist_raw",		1,		&BehaviorGen::callbackGetTwistRaw, 	this);
+	sub_twist_cmd = nh.subscribe("/twist_cmd",		1,		&BehaviorGen::callbackGetTwistCMD, 	this);
+	//sub_ctrl_cmd = nh.subscribe("/ctrl_cmd",		1,		&BehaviorGen::callbackGetCommandCMD, 	this);
+}
+
+BehaviorGen::~BehaviorGen()
+{
+	UtilityHNS::DataRW::WriteLogData(UtilityHNS::UtilityH::GetHomeDirectory()+UtilityHNS::DataRW::LoggingMainfolderName+UtilityHNS::DataRW::StatesLogFolderName, "MainLog",
+				"time,dt, Behavior_i, Behavior_str, RollOuts_n, Blocked_i, Central_i, Selected_i, StopSign_id, Light_id, Stop_Dist, Follow_Dist, Follow_Vel,"
+				"Target_Vel, PID_Vel, T_cmd_Vel, C_cmd_Vel, Vel, Steer, X, Y, Z, Theta,"
+				, m_LogData);
+}
+
+void BehaviorGen::UpdatePlanningParams(ros::NodeHandle& _nh)
+{
+	_nh.getParam("/op_common_params/enableSwerving", m_PlanningParams.enableSwerving);
+	if(m_PlanningParams.enableSwerving)
```

Comment:
If I understand it right, if "enableSwerving" is set, "enableFollowing" would be set and then in "StateMachine" the state would be updated to "Following" or "Obstacle Avoidance" depending on whether obstacle is fully blocking or not. Would it be more reasonable to incorporate this part of logic into StateMachine instead of doing it in different places ? Like  "BehaviorGen", "TrajectoryGen", "MotionPrediction", "TrajectoryEval" separately?

Reply:
It is implemented as state machine in behavior_selector , the condition just control if certain states are included in the loop or not, and because "following" and "avoidance" is related conditionally - following could be used without avoidance, but when avoidance is activated following must be activated by default-